
# Analyze Debtor's Default Risk

Our job is to prepare a report for the credit division of a bank. We find out the effect of a customer's marital status and the number of children he has on the probability of default in repayment of the loan. The bank already has some data regarding the creditworthiness of customers.

Our report will be considered when making **credit assessment** for prospective customers. **Credit scoring** is used to evaluate the ability of potential borrowers to repay their loans.

## Open the data *file* and read the general information.

In [1]:
import pandas as pd

df = pd.read_csv('/datasets/credit_scoring_eng.csv')

## Data Explore

**Data Description**
- `children` - the number of children in the family
- `days_employed` - the customer's work experience in days
- `dob_years` - customer's age in years
- `education` - level of customer education
- `education_id` - identifier for the customer's education level
- `family_status` - identifier for the customer's marital status
- `family_status_id` - marital status identifier
- `gender` - the customer's gender
- `income_type` - type of work
- `debt` - whether the customer has a loan payment debt
- `total_income` - monthly income
- `purpose` - the purpose of getting a loan

In [2]:
df.shape

(21525, 12)

In [3]:
df.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   
15         1    -972.364419         26  secondary education             1   
16         0   -1719.934226         35  secondary education             1   
17         0   -2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0  -10038.818549         48  SECONDARY EDUCATION             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate  
15     18691.345         construction of own property  
16   

<div class="alert alert-block alert-info">Based on the sample table above showing 20 rows, here are some issues that might require further investigation.
    
- Column 'days_employed' still has a lot of negative values. There is 1 row whose value is 'NaN'. And all rows are still float values.
- Column 'education' has duplicate values, for example it contains the same meaning but different writing (Uppercase & Lowercase).
- Column 'total_income' has 1 row which has the value 'NaN' and is aligned with 'NaN' in 'days_employed'. All rows are still float values.<a class="tocSkip"></a></div>

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-block alert-info">Based on the info() method above, there are missing values ​​in the 'days_employed' and 'total_income' columns. The number of missing values ​​in both columns is the same.</br></br><a class="tocSkip">

In [5]:
df[df['days_employed'].isna()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

<div class="alert alert-block alert-info">Based on the filter with the 'days_employed' column with missing values, it appears that the missing values ​​are symmetrical in the 'total_income' column. However this still requires further investigation.<a class="tocSkip"></a></div>

In [6]:
df.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Conclusion**

<div class="alert alert-block alert-info">The number of table rows that have been filtered matches the number of missing values. It means that there is a pattern or not random. <a class="tocSkip"></a></div>

<div class="alert alert-block alert-info">The number of missing rows is 21525 - 19351 = 2174 which is about 10%.<a class="tocSkip"></a></div>

<div class="alert alert-block alert-info">
    
- Columns 'days_employed' and 'total_income' are quantitative variables.
- Missing values ​​for these variables can be filled in with representative values.
- The representative value can use the median or mean of each column. <a class="tocSkip"></a></div>

In [7]:
df_isnull = df[df['days_employed'].isnull()]
df_isnull

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [8]:
df_isnull.describe()

children  days_employed    dob_years  education_id  \
count  2174.000000            0.0  2174.000000   2174.000000   
mean      0.552438            NaN    43.632015      0.800828   
std       1.469356            NaN    12.531481      0.530157   
min      -1.000000            NaN     0.000000      0.000000   
25%       0.000000            NaN    34.000000      0.250000   
50%       0.000000            NaN    43.000000      1.000000   
75%       1.000000            NaN    54.000000      1.000000   
max      20.000000            NaN    73.000000      3.000000   

       family_status_id         debt  total_income  
count       2174.000000  2174.000000           0.0  
mean           0.975161     0.078197           NaN  
std            1.418220     0.268543           NaN  
min            0.000000     0.000000           NaN  
25%            0.000000     0.000000           NaN  
50%            0.000000     0.000000           NaN  
75%            1.000000     0.000000           NaN  
max            4.000000     1.000000           NaN

**Possible causes of missing values ​​in data**

In [9]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

**Conclusion**
<div class="alert alert-block alert-info">Based on the table df_notnull (table with non-null filter) and df(original table) the distribution is similar. Meaning that if we delete/drop a null row it shouldn't really matter.<a class="tocSkip"></a></div>

In [10]:
print("Calculating missing values ​​'days_employed' based on 'gender'")
print(df[df['days_employed'].isnull()]['gender'].value_counts())

print()
print("Calculating missing value 'total_income' based on 'gender'")
print(df[df['total_income'].isnull()]['gender'].value_counts())

Calculating missing values ​​'days_employed' based on 'gender'
F    1484
M     690
Name: gender, dtype: int64

Calculating missing value 'total_income' based on 'gender'
F    1484
M     690
Name: gender, dtype: int64


**Conclusion**
<div class="alert alert-block alert-info">From the calculation above which results in the number of missing values ​​based on gender, it can be concluded that the loss of values ​​does not occur randomly.<a class="tocSkip"></a></div>

<div class="alert alert-block alert-info">All the temporary conclusions above are based on the following:
    
- Display first 20 Rows on Dataframe.
- Use of the info() and shape methods for general dataframe information.
- Use isna(), isnull() for table filtering with missing values.
- Use of describe() for descriptive statistics of Dataframes.

We need to fix those missing values ​​and improve the Dataframe further by doing Data Transformation and Data Categorization.<a class="tocSkip"></a></div>

## Data transformation

In [11]:
df['education'].sort_values().unique()

array(["BACHELOR'S DEGREE", "Bachelor's Degree", 'GRADUATE DEGREE',
       'Graduate Degree', 'PRIMARY EDUCATION', 'Primary Education',
       'SECONDARY EDUCATION', 'SOME COLLEGE', 'Secondary Education',
       'Some College', "bachelor's degree", 'graduate degree',
       'primary education', 'secondary education', 'some college'],
      dtype=object)

In [12]:
df['education'] = df['education'].str.lower()

In [13]:
df['education'].value_counts()

secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

In [14]:
df['children'].sort_values().unique()

array([-1,  0,  1,  2,  3,  4,  5, 20])

<div class="alert alert-block alert-info">Here are some explanations for the odd thing to do with values ​​in the 'children' column.
    
- There is a value of -1, this might be a typo, we can change the value to 1 using the .replace() method
- Maybe we see something strange in the value of 20, is it true that the customer has 20 children? However, based on studies, a woman can have a maximum of 15-30 children for 15 pregnancies. 
    <a class="tocSkip"></a></div>

In [15]:
df['children'] = df['children'].replace(-1, 1)

In [16]:
df['children'].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

In [17]:
df['days_employed'].sort_values().unique()

array([-18388.94990057, -17615.56326563, -16593.47281726, ...,
       401715.81174889, 401755.40047533,             nan])

In [18]:
days_employed_null = df['days_employed'].isnull().sum()/df.shape[0]
print('Jumlah percentage untuk nilai yang hilang pada kolom days_employed adalah {:.2%}'.format(days_employed_null))

Jumlah percentage untuk nilai yang hilang pada kolom days_employed adalah 10.10%


In [19]:
df['days_employed'] = abs(df['days_employed'])

<div class="alert alert-block alert-info">During the temporary conclusion in the Data Exploration section, several problems have been explained in the 'days_employed' column:
    
- Some data are negative, this needs to be fixed by using the abs() function.
- Rows with 'NaN' values ​​will be corrected in the 'Working with missing values' section.
- All rows are still of type float, this needs to be fixed using astype() after fixing Value 'NaN'..<a class="tocSkip"></a></div>

In [20]:
df['days_employed'].head(10)

0      8437.673028
1      4024.803754
2      5623.422610
3      4124.747207
4    340266.072047
5       926.185831
6      2879.202052
7       152.779569
8      6929.865299
9      2188.756445
Name: days_employed, dtype: float64

In [21]:
df['dob_years'].max()

75

In [22]:
340266.072047 / 365

932.2358138273972

In [23]:
(df['days_employed'] / 365).describe()

count    19351.000000
mean       183.328024
std        380.906522
min          0.066141
25%          2.539751
50%          6.011563
75%         15.172281
max       1100.699727
Name: days_employed, dtype: float64

In [24]:
df['days_employed'].describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

In [25]:
df.sort_values(by='days_employed' , ascending=False)

children  days_employed  dob_years            education  education_id  \
6954          0  401755.400475         56  secondary education             1   
10006         0  401715.811749         69    bachelor's degree             0   
7664          1  401675.093434         61  secondary education             1   
2156          0  401674.466633         60  secondary education             1   
7794          0  401663.850046         61  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  secondary education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    bachelor's degree             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
6954     widow / widower                 2      F     retiree     0   
10006          unmarried                 4      F     retiree     0   
7664             married                 0      F     retiree     0   
2156             married                 0      M     retiree     0   
7794   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21489            married                 0      M    business     0   
21495  civil partnership                 1      F    employee     0   
21497            married                 0      F    business     0   
21502            married                 0      F    employee     0   
21510            married                 0      F    employee     0   

       total_income                 purpose  
6954      28204.551      housing renovation  
10006      9182.441    getting an education  
7664      20194.323    housing transactions  
2156      52063.316                    cars  
7794       7725.831        wedding ceremony  
...             ...                     ...  
21489           NaN       purchase of a car  
21495           NaN        wedding ceremony  
21497           NaN     building a property  
21502           NaN  building a real estate  
21510           NaN            car purchase  

[21525 rows x 12 columns]

In [26]:
median_days_employed = df['days_employed'].median()
median_days_employed

2194.220566878695

In [27]:
dob_days_max = df['dob_years'].max() * 365
dob_days_max

27375

In [28]:
df.loc[df['days_employed'] >= 27375, 'days_employed'] = median_days_employed
df.sort_values(by='days_employed', ascending=False).head(10)

children  days_employed  dob_years            education  education_id  \
16335         1   18388.949901         61  secondary education             1   
4299          0   17615.563266         61  secondary education             1   
7329          0   16593.472817         60    bachelor's degree             0   
17838         0   16264.699501         59  secondary education             1   
16825         0   16119.687737         64  secondary education             1   
3974          0   15835.725775         64  secondary education             1   
1539          0   15785.678893         59    bachelor's degree             0   
4321          0   15773.061335         61  secondary education             1   
7731          0   15618.063786         64  secondary education             1   
15675         0   15410.040779         65    bachelor's degree             0   

           family_status  family_status_id gender income_type  debt  \
16335            married                 0      F    employee     0   
4299             married                 0      F    business     0   
7329             married                 0      F    employee     0   
17838            married                 0      F    employee     0   
16825            married                 0      F    employee     0   
3974   civil partnership                 1      F    business     0   
1539           unmarried                 4      F    employee     0   
4321   civil partnership                 1      F    employee     0   
7731             married                 0      F    business     0   
15675            married                 0      F    employee     0   

       total_income                                   purpose  
16335     29788.629                  real estate transactions  
4299      19609.719                     purchase of the house  
7329      19951.655                       going to university  
17838      8198.235                              to buy a car  
16825     14644.430               buy residential real estate  
3974      15497.365                         to have a wedding  
1539      19130.216  transactions with commercial real estate  
4321      32938.974                          wedding ceremony  
7731      47444.057                      university education  
15675     30208.011               buy residential real estate

In [29]:
df['days_employed'].max() / 365

50.38068465909146

<div class="alert alert-block alert-info">There are customers who are worth 0 years.<a class="tocSkip"></a></div>

In [30]:
dob_years_count = df['dob_years'].count()
dob_years_zero = df[df['dob_years'] == 0]
dob_years_zero = dob_years_zero.shape[0]

dob_years_zero_percent = dob_years_zero / dob_years_count
dob_years_zero_percent
print('The total percentage for customers aged 0 years is {:.2%}'.format(dob_years_zero_percent))



The total percentage for customers aged 0 years is 0.47%


<div class="alert alert-block alert-info">Because our goal is to find out the effect of a customer's marital status and number of children on the probability of default on a loan, the percentage for customers aged 0 years is only 0.47%, which doesn't really matter. It is also possible that this happened because of a typing error. So for now we leave it alone, do not delete these lines.<a class="tocSkip"></a></div>

In [31]:
df['family_status'].sort_values().unique()

array(['civil partnership', 'divorced', 'married', 'unmarried',
       'widow / widower'], dtype=object)

In [32]:
df['gender'].sort_values().unique()

array(['F', 'M', 'XNA'], dtype=object)

In [33]:
df = df.loc[df['gender'] != 'XNA']

In [34]:
df['gender'].value_counts()

F    14236
M     7288
Name: gender, dtype: int64

In [35]:
df['income_type'].sort_values().unique()

array(['business', 'civil servant', 'employee', 'entrepreneur',
       'paternity / maternity leave', 'retiree', 'student', 'unemployed'],
      dtype=object)

In [36]:
df[ (df['income_type'] == 'student') | (df['income_type'] == 'unemployed')]

children  days_employed  dob_years            education  education_id  \
3133          1    2194.220567         31  secondary education             1   
9410          0     578.751554         22    bachelor's degree             0   
14798         0    2194.220567         45    bachelor's degree             0   

           family_status  family_status_id gender income_type  debt  \
3133             married                 0      M  unemployed     1   
9410           unmarried                 4      M     student     0   
14798  civil partnership                 1      F  unemployed     0   

       total_income                          purpose  
3133       9593.119  buying property for renting out  
9410      15712.260     construction of own property  
14798     32435.602               housing renovation

<div class="alert alert-block alert-info">Although there are some customers whose 'income_type' is student and unemployed, some of them have debt = 0 (no loan repayments). So for now there is nothing to fix in the column 'income_type'.<a class="tocSkip"></a></div>

In [37]:
df[df.duplicated()]

children  days_employed  dob_years            education  education_id  \
2849          0            NaN         41  secondary education             1   
3290          0            NaN         58  secondary education             1   
4182          1            NaN         34    bachelor's degree             0   
4851          0            NaN         60  secondary education             1   
5557          0            NaN         58  secondary education             1   
...         ...            ...        ...                  ...           ...   
20702         0            NaN         64  secondary education             1   
21032         0            NaN         60  secondary education             1   
21132         0            NaN         47  secondary education             1   
21281         1            NaN         30    bachelor's degree             0   
21415         0            NaN         54  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
2849             married                 0      F    employee     0   
3290   civil partnership                 1      F     retiree     0   
4182   civil partnership                 1      F    employee     0   
4851   civil partnership                 1      F     retiree     0   
5557   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
20702            married                 0      F     retiree     0   
21032            married                 0      F     retiree     0   
21132            married                 0      F    employee     0   
21281            married                 0      F    employee     0   
21415            married                 0      F     retiree     0   

       total_income                              purpose  
2849            NaN  purchase of the house for my family  
3290            NaN                    to have a wedding  
4182            NaN                     wedding ceremony  
4851            NaN                     wedding ceremony  
5557            NaN                    to have a wedding  
...             ...                                  ...  
20702           NaN              supplementary education  
21032           NaN                   to become educated  
21132           NaN                   housing renovation  
21281           NaN           buy commercial real estate  
21415           NaN                 housing transactions  

[71 rows x 12 columns]

In [38]:
df_credit_analysis = df.drop_duplicates().reset_index(drop=True)

In [39]:
df_credit_analysis[df_credit_analysis.duplicated()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [40]:
df.shape

(21524, 12)

In [41]:
df_credit_analysis.shape

(21453, 12)

In [42]:
df_shape1 = df.shape[0]
df_shape2 = df_credit_analysis.shape[0]
df_shape_percent =   1 - (int(df_shape2) * 1) / int(df_shape1)
df_shape_percent

0.003298643374837429

In [43]:
print('Total percentage for changes in dataset size is {:.2%}'.format(df_shape_percent))

Total percentage for changes in dataset size is 0.33%


<div class="alert alert-block alert-info">
The number of percentage changes with deleted duplicates is still very small, so it is necessary to correct the missing values<a class="tocSkip"></a></div>

# Fill missing values

In [44]:
df_credit_analysis.describe()

children  days_employed     dob_years  education_id  \
count  21453.000000   19350.000000  21453.000000  21453.000000   
mean       0.544353    2324.744324     43.272130      0.817042   
std        1.381753    2090.015644     12.570427      0.548628   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000     926.990457     33.000000      1.000000   
50%        0.000000    2194.218768     42.000000      1.000000   
75%        1.000000    2747.486276     53.000000      1.000000   
max       20.000000   18388.949901     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21453.000000  21453.000000   19350.000000  
mean           0.973896      0.081154   26787.266688  
std            1.421601      0.273078   16475.822926  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16486.515250  
50%            0.000000      0.000000   23201.873500  
75%            1.000000      0.000000   32547.910750  
max            4.000000      1.000000  362496.645000

### Fixed missing value in `total_income`

<div class="alert alert-block alert-info">As previously concluded, the 'days_employed' and 'total_income' columns are missing values. With a total value of around 10%, it is actually possible to delete rows with missing values, but you can also fill in missing values ​​with representative values.<a class="tocSkip"></a></div>


In [45]:
df_credit_analysis['dob_years'].sort_values().unique()

array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75])

In [46]:
def dob_years_group(age):
    if age < 0 or pd.isna(age):
        return 'NA'
    elif age < 10:
        return '0-9'
    elif age < 20:
        return '10-19' 
    elif age < 30:
        return '20-29'
    elif age < 40:
        return '30-39'
    elif age < 50:
        return '40-49'
    elif age < 60:
        return '50-59'
    elif age < 70:
        return '60-69'
    else:
        return '70+'

In [47]:
dob_years_group(69)

'60-69'

In [48]:
df_credit_analysis['dob_years_group'] = df_credit_analysis['dob_years'].apply(dob_years_group)

In [49]:
df_credit_analysis.head(10)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    2194.220567         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose dob_years_group  
0                purchase of the house           40-49  
1                         car purchase           30-39  
2                purchase of the house           30-39  
3              supplementary education           30-39  
4                    to have a wedding           50-59  
5                purchase of the house           20-29  
6                 housing transactions           40-49  
7                            education           50-59  
8                     having a wedding           30-39  
9  purchase of the house for my family           40-49

In [50]:
df_credit_analysis.pivot_table(index='dob_years_group', values='total_income', aggfunc=['median', 'mean'])

median          mean
                total_income  total_income
dob_years_group                           
0-9               24387.0700  25334.072890
10-19             14934.9010  16993.942462
20-29             22798.6650  25570.172966
30-39             24667.5280  28312.479963
40-49             24764.2290  28551.375635
50-59             22203.0745  25811.700327
60-69             19817.4400  23242.812818
70+               18751.3240  20125.658331

In [51]:
df_credit_analysis_notnull = df_credit_analysis[df_credit_analysis['total_income'].notnull()]
df_credit_analysis_notnull
df_credit_analysis_notnull.info()

print()
print("Informasi statistik deskriptif pada 'df_credit_analysis_notnull' dengan kolom 'total_income' :")
df_credit_analysis_notnull['total_income'].describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19350 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          19350 non-null  int64  
 1   days_employed     19350 non-null  float64
 2   dob_years         19350 non-null  int64  
 3   education         19350 non-null  object 
 4   education_id      19350 non-null  int64  
 5   family_status     19350 non-null  object 
 6   family_status_id  19350 non-null  int64  
 7   gender            19350 non-null  object 
 8   income_type       19350 non-null  object 
 9   debt              19350 non-null  int64  
 10  total_income      19350 non-null  float64
 11  purpose           19350 non-null  object 
 12  dob_years_group   19350 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB

Informasi statistik deskriptif pada 'df_credit_analysis_notnull' dengan kolom 'total_income' :


count     19350.000000
mean      26787.266688
std       16475.822926
min        3306.762000
25%       16486.515250
50%       23201.873500
75%       32547.910750
max      362496.645000
Name: total_income, dtype: float64

In [52]:
df_credit_analysis_notnull['total_income'].describe()['mean']

26787.26668764858

In [53]:
df_credit_analysis_notnull['total_income'].describe()['50%']

23201.8735

In [54]:
df_credit_analysis_notnull.groupby('gender')['total_income'].mean()

gender
F    24655.604757
M    30907.144369
Name: total_income, dtype: float64

In [55]:
df_credit_analysis_notnull.groupby('family_status')['total_income'].mean()

family_status
civil partnership    26692.840381
divorced             27189.354550
married              27041.784689
unmarried            26934.069805
widow / widower      22984.208556
Name: total_income, dtype: float64

In [56]:
df_credit_analysis_notnull.groupby('gender')['total_income'].median()

gender
F    21464.845
M    26834.295
Name: total_income, dtype: float64

In [57]:
df_credit_analysis_notnull.groupby('family_status')['total_income'].median()

family_status
civil partnership    23185.477
divorced             23515.096
married              23389.540
unmarried            23149.028
widow / widower      20514.190
Name: total_income, dtype: float64

<div class="alert alert-block alert-info">Based on the describe()[50%] method, the median is 23201, and based on the groupby('family_status') method, the median appears around 23K. So we use the median because if we use the mean the range is too much.<a class="tocSkip"></a></div>


In [58]:
df_credit_analysis.groupby(['income_type'])['total_income'].median()

income_type
business                       27571.0825
civil servant                  24071.6695
employee                       22815.1035
entrepreneur                   79866.1030
paternity / maternity leave     8612.6610
retiree                        18962.3180
student                        15712.2600
unemployed                     21014.3605
Name: total_income, dtype: float64

In [59]:
business_median = df_credit_analysis['income_type'] == 'business'
df_credit_analysis.loc[business_median, 'total_income'] = df_credit_analysis.loc[business_median, 'total_income'].fillna(27571)

civil_median = df_credit_analysis['income_type'] == 'civil servant'
df_credit_analysis.loc[civil_median, 'total_income'] = df_credit_analysis.loc[civil_median, 'total_income'].fillna(24071)

employee_median = df_credit_analysis['income_type'] == 'employee'
df_credit_analysis.loc[employee_median, 'total_income'] = df_credit_analysis.loc[employee_median, 'total_income'].fillna(22815)

entrepreneur_median = df_credit_analysis['income_type'] == 'entrepreneur'
df_credit_analysis.loc[entrepreneur_median, 'total_income'] = df_credit_analysis.loc[entrepreneur_median, 'total_income'].fillna(79866)

leave_median = df_credit_analysis['income_type'] == 'paternity / maternity leave'
df_credit_analysis.loc[leave_median, 'total_income'] = df_credit_analysis.loc[leave_median, 'total_income'].fillna(8612)

retiree_median = df_credit_analysis['income_type'] == 'retiree'
df_credit_analysis.loc[retiree_median, 'total_income'] = df_credit_analysis.loc[retiree_median, 'total_income'].fillna(18962)

student_median = df_credit_analysis['income_type'] == 'student'
df_credit_analysis.loc[student_median, 'total_income'] = df_credit_analysis.loc[student_median, 'total_income'].fillna(15712)

unemployed_median = df_credit_analysis['income_type'] == 'unemployed'
df_credit_analysis.loc[unemployed_median, 'total_income'] = df_credit_analysis.loc[unemployed_median, 'total_income'].fillna(21014)


In [60]:
df_credit_analysis[df_credit_analysis['total_income'].isnull()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, dob_years_group]
Index: []

In [61]:
df_credit_analysis[df_credit_analysis['total_income'].isna()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, dob_years_group]
Index: []

In [62]:
df_credit_analysis.describe()

children  days_employed     dob_years  education_id  \
count  21453.000000   19350.000000  21453.000000  21453.000000   
mean       0.544353    2324.744324     43.272130      0.817042   
std        1.381753    2090.015644     12.570427      0.548628   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000     926.990457     33.000000      1.000000   
50%        0.000000    2194.218768     42.000000      1.000000   
75%        1.000000    2747.486276     53.000000      1.000000   
max       20.000000   18388.949901     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21453.000000  21453.000000   21453.000000  
mean           0.973896      0.081154   26450.763269  
std            1.421601      0.273078   15710.271920  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   17219.352000  
50%            0.000000      0.000000   22815.000000  
75%            1.000000      0.000000   31331.009000  
max            4.000000      1.000000  362496.645000

In [63]:
df_credit_analysis.groupby(['income_type'])['total_income'].mean()

income_type
business                       31911.522860
civil servant                  27018.028972
employee                       25530.671113
entrepreneur                   79866.051500
paternity / maternity leave     8612.661000
retiree                        21640.143712
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64

In [64]:
df_credit_analysis.groupby(['income_type'])['total_income'].median()

income_type
business                       27571.0000
civil servant                  24071.0000
employee                       22815.0000
entrepreneur                   79866.0515
paternity / maternity leave     8612.6610
retiree                        18962.0000
student                        15712.2600
unemployed                     21014.3605
Name: total_income, dtype: float64

In [65]:
df_credit_analysis[df_credit_analysis['days_employed'].isna()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21417         2            NaN         47  secondary education             1   
21423         1            NaN         50  secondary education             1   
21425         0            NaN         48    bachelor's degree             0   
21430         1            NaN         42  secondary education             1   
21438         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21417            married                 0      M       business     0   
21423  civil partnership                 1      F       employee     0   
21425            married                 0      F       business     0   
21430            married                 0      F       employee     0   
21438            married                 0      F       employee     0   

       total_income                   purpose dob_years_group  
12          18962.0         to have a wedding           60-69  
26          24071.0                 education           40-49  
29          18962.0    building a real estate           60-69  
41          24071.0  second-hand car purchase           50-59  
55          18962.0         to have a wedding           50-59  
...             ...                       ...             ...  
21417       27571.0         purchase of a car           40-49  
21423       22815.0          wedding ceremony           50-59  
21425       27571.0       building a property           40-49  
21430       22815.0    building a real estate           40-49  
21438       22815.0              car purchase           20-29  

[2103 rows x 13 columns]

In [66]:
df_credit_analysis.isnull().sum()

children               0
days_employed       2103
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dob_years_group        0
dtype: int64

In [67]:
df_credit_analysis['total_income'] = df_credit_analysis['total_income'].astype('int')

In [68]:
df_credit_analysis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21453 non-null  int64  
 1   days_employed     19350 non-null  float64
 2   dob_years         21453 non-null  int64  
 3   education         21453 non-null  object 
 4   education_id      21453 non-null  int64  
 5   family_status     21453 non-null  object 
 6   family_status_id  21453 non-null  int64  
 7   gender            21453 non-null  object 
 8   income_type       21453 non-null  object 
 9   debt              21453 non-null  int64  
 10  total_income      21453 non-null  int64  
 11  purpose           21453 non-null  object 
 12  dob_years_group   21453 non-null  object 
dtypes: float64(1), int64(6), object(6)
memory usage: 2.1+ MB


###  Fixed the value in `days_employed`

In [69]:
# The median distribution of `days_employed` based on the parameters we identified
df_credit_analysis.groupby(['income_type'])['days_employed'].median()



income_type
business                       1546.333214
civil servant                  2689.368353
employee                       1574.202821
entrepreneur                    520.848083
paternity / maternity leave    3296.759962
retiree                        2194.220567
student                         578.751554
unemployed                     2194.220567
Name: days_employed, dtype: float64

In [70]:
# Average distribution of `days_employed` based on the parameters you identified
df_credit_analysis.groupby(['income_type'])['days_employed'].mean()

income_type
business                       2111.470404
civil servant                  3399.896902
employee                       2326.499216
entrepreneur                    520.848083
paternity / maternity leave    3296.759962
retiree                        2194.220567
student                         578.751554
unemployed                     2194.220567
Name: days_employed, dtype: float64

In [71]:
df_credit_analysis.describe()

children  days_employed     dob_years  education_id  \
count  21453.000000   19350.000000  21453.000000  21453.000000   
mean       0.544353    2324.744324     43.272130      0.817042   
std        1.381753    2090.015644     12.570427      0.548628   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000     926.990457     33.000000      1.000000   
50%        0.000000    2194.218768     42.000000      1.000000   
75%        1.000000    2747.486276     53.000000      1.000000   
max       20.000000   18388.949901     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21453.000000  21453.000000   21453.000000  
mean           0.973896      0.081154   26450.314921  
std            1.421601      0.273078   15710.264006  
min            0.000000      0.000000    3306.000000  
25%            0.000000      0.000000   17219.000000  
50%            0.000000      0.000000   22815.000000  
75%            1.000000      0.000000   31331.000000  
max            4.000000      1.000000  362496.000000

<div class="alert alert-block alert-info">Before replacing 'days_employed' with a value > 27375 days or 75 years, based on the describe() method the median is ~2194, and the mean is 66918. With conditions like this it means that there is a significant otulier, it is better to use the median value.<a class="tocSkip"></a></div>

In [72]:
business_median_days = df_credit_analysis['income_type'] == 'business'
df_credit_analysis.loc[business_median_days, 'days_employed'] = df_credit_analysis.loc[business_median_days, 'days_employed'].fillna(1546)

civil_median_days = df_credit_analysis['income_type'] == 'civil servant'
df_credit_analysis.loc[civil_median_days, 'days_employed'] = df_credit_analysis.loc[civil_median_days, 'days_employed'].fillna(2689)

employee_median_days = df_credit_analysis['income_type'] == 'employee'
df_credit_analysis.loc[employee_median_days, 'days_employed'] = df_credit_analysis.loc[employee_median_days, 'days_employed'].fillna(1574)

entrepreneur_median_days = df_credit_analysis['income_type'] == 'entrepreneur'
df_credit_analysis.loc[entrepreneur_median_days, 'days_employed'] = df_credit_analysis.loc[entrepreneur_median_days, 'days_employed'].fillna(520)

leave_median_days = df_credit_analysis['income_type'] == 'paternity / maternity leave'
df_credit_analysis.loc[leave_median_days, 'days_employed'] = df_credit_analysis.loc[leave_median_days, 'days_employed'].fillna(3296)

retiree_median_days = df_credit_analysis['income_type'] == 'retiree'
df_credit_analysis.loc[retiree_median_days, 'days_employed'] = df_credit_analysis.loc[retiree_median_days, 'days_employed'].fillna(2194)

student_median_days = df_credit_analysis['income_type'] == 'student'
df_credit_analysis.loc[student_median_days, 'days_employed'] = df_credit_analysis.loc[student_median_days, 'days_employed'].fillna(1578)

unemployed_median_days = df_credit_analysis['income_type'] == 'unemployed'
df_credit_analysis.loc[unemployed_median_days, 'days_employed'] = df_credit_analysis.loc[unemployed_median_days, 'days_employed'].fillna(2194)


df_credit_analysis.loc[business_median_days, 'days_employed'].isna().sum()
df_credit_analysis[df_credit_analysis['days_employed'] == 1546]

children  days_employed  dob_years            education  education_id  \
65            0         1546.0         21  secondary education             1   
94            1         1546.0         34    bachelor's degree             0   
121           0         1546.0         29    bachelor's degree             0   
135           0         1546.0         27  secondary education             1   
174           0         1546.0         55    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
21319        20         1546.0         53  secondary education             1   
21320         0         1546.0         52  secondary education             1   
21336         1         1546.0         36  secondary education             1   
21417         2         1546.0         47  secondary education             1   
21425         0         1546.0         48    bachelor's degree             0   

           family_status  family_status_id gender income_type  debt  \
65             unmarried                 4      M    business     0   
94     civil partnership                 1      F    business     0   
121              married                 0      F    business     0   
135              married                 0      M    business     0   
174      widow / widower                 2      F    business     0   
...                  ...               ...    ...         ...   ...   
21319            married                 0      M    business     0   
21320            married                 0      F    business     0   
21336            married                 0      F    business     0   
21417            married                 0      M    business     0   
21425            married                 0      F    business     0   

       total_income                                   purpose dob_years_group  
65            27571  transactions with commercial real estate           20-29  
94            27571                          having a wedding           30-39  
121           27571                                       car           20-29  
135           27571                                   housing           20-29  
174           27571                              to own a car           50-59  
...             ...                                       ...             ...  
21319         27571               buy residential real estate           50-59  
21320         27571       purchase of the house for my family           50-59  
21336         27571                    building a real estate           30-39  
21417         27571                         purchase of a car           40-49  
21425         27571                       building a property           40-49  

[501 rows x 13 columns]

In [73]:
#Check if our function can work
df_credit_analysis.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dob_years_group     0
dtype: int64

In [74]:
# Check entries in all columns - make sure we correct all missing values
df_credit_analysis.isnull().sum()
df_credit_analysis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21453 non-null  int64  
 1   days_employed     21453 non-null  float64
 2   dob_years         21453 non-null  int64  
 3   education         21453 non-null  object 
 4   education_id      21453 non-null  int64  
 5   family_status     21453 non-null  object 
 6   family_status_id  21453 non-null  int64  
 7   gender            21453 non-null  object 
 8   income_type       21453 non-null  object 
 9   debt              21453 non-null  int64  
 10  total_income      21453 non-null  int64  
 11  purpose           21453 non-null  object 
 12  dob_years_group   21453 non-null  object 
dtypes: float64(1), int64(6), object(6)
memory usage: 2.1+ MB


In [75]:
df_credit_analysis['days_employed'] = df_credit_analysis['days_employed'].astype('int')

In [76]:
df_credit_analysis.info()
df_credit_analysis.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21453 non-null  int64 
 1   days_employed     21453 non-null  int64 
 2   dob_years         21453 non-null  int64 
 3   education         21453 non-null  object
 4   education_id      21453 non-null  int64 
 5   family_status     21453 non-null  object
 6   family_status_id  21453 non-null  int64 
 7   gender            21453 non-null  object
 8   income_type       21453 non-null  object
 9   debt              21453 non-null  int64 
 10  total_income      21453 non-null  int64 
 11  purpose           21453 non-null  object
 12  dob_years_group   21453 non-null  object
dtypes: int64(7), object(6)
memory usage: 2.1+ MB


children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0           2194         53  secondary education             1   
5         0            926         27    bachelor's degree             0   
6         0           2879         43    bachelor's degree             0   
7         0            152         50  secondary education             1   
8         2           6929         35    bachelor's degree             0   
9         0           2188         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   
5  civil partnership                 1      M    business     0         40922   
6            married                 0      F    business     0         38484   
7            married                 0      M    employee     0         21731   
8  civil partnership                 1      F    employee     0         15337   
9            married                 0      M    employee     0         23108   

                               purpose dob_years_group  
0                purchase of the house           40-49  
1                         car purchase           30-39  
2                purchase of the house           30-39  
3              supplementary education           30-39  
4                    to have a wedding           50-59  
5                purchase of the house           20-29  
6                 housing transactions           40-49  
7                            education           50-59  
8                     having a wedding           30-39  
9  purchase of the house for my family           40-49

In [77]:
df_credit_analysis.describe()

children  days_employed     dob_years  education_id  \
count  21453.000000   21453.000000  21453.000000  21453.000000   
mean       0.544353    2268.733557     43.272130      0.817042   
std        1.381753    1995.103503     12.570427      0.548628   
min        0.000000      24.000000      0.000000      0.000000   
25%        0.000000    1023.000000     33.000000      1.000000   
50%        0.000000    1996.000000     42.000000      1.000000   
75%        1.000000    2575.000000     53.000000      1.000000   
max       20.000000   18388.000000     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21453.000000  21453.000000   21453.000000  
mean           0.973896      0.081154   26450.314921  
std            1.421601      0.273078   15710.264006  
min            0.000000      0.000000    3306.000000  
25%            0.000000      0.000000   17219.000000  
50%            0.000000      0.000000   22815.000000  
75%            1.000000      0.000000   31331.000000  
max            4.000000      1.000000  362496.000000

## Data Categorization

In [78]:
# Displays the data values we selected for categorization
df_credit_analysis[['total_income', 'purpose']]


total_income                  purpose
0             40620    purchase of the house
1             17932             car purchase
2             23341    purchase of the house
3             42820  supplementary education
4             25378        to have a wedding
...             ...                      ...
21448         35966     housing transactions
21449         24959        purchase of a car
21450         14347                 property
21451         39054        buying my own car
21452         13127             to buy a car

[21453 rows x 2 columns]

In [79]:
# Check for unique values
df_credit_analysis['purpose'].sort_values().unique()

array(['building a property', 'building a real estate',
       'buy commercial real estate', 'buy real estate',
       'buy residential real estate', 'buying a second-hand car',
       'buying my own car', 'buying property for renting out', 'car',
       'car purchase', 'cars', 'construction of own property',
       'education', 'getting an education', 'getting higher education',
       'going to university', 'having a wedding', 'housing',
       'housing renovation', 'housing transactions', 'profile education',
       'property', 'purchase of a car', 'purchase of my own house',
       'purchase of the house', 'purchase of the house for my family',
       'real estate transactions', 'second-hand car purchase',
       'supplementary education', 'to become educated', 'to buy a car',
       'to get a supplementary education', 'to have a wedding',
       'to own a car', 'transactions with commercial real estate',
       'transactions with my real estate', 'university education',
       'we

In [80]:
# function to categorize data by common topics
def purpose_group(purpose_list):
    if purpose_list in ['supplementary education', 'university education', 'education', 'to get a supplementary education', 'getting an education', 
     'getting higher education', 'profile education', 'to become educated', 'going to university']:
        return 'education'
    elif purpose_list in ['wedding', 'to have a wedding', 'having a wedding', 'wedding ceremony']:
        return 'wedding'
    elif purpose_list in ['real estate transactions', 'buy commercial real estate', 'building a real estate', 'transactions with commercial real estate', 
     'transactions with my real estate', 'buy real estate', 'buy residential real estate']:
        return 'real_estate' 
    elif purpose_list in ['buying property for renting out', 'construction of own property', 'property', 'building a property']:
        return 'property'
    elif purpose_list in  ['purchase of the house', 'housing', 'housing transactions', 'purchase of my own house', 
     'purchase of the house for my family', 'housing renovation']:
        return 'house'
    elif purpose_list in ['car', 'buying my own car', 'to buy a car', 'buying a second-hand car', 'second-hand car purchase', 'to own a car', 
     'cars', 'car purchase', 'purchase of a car']:
        return 'car'



In [81]:
df_credit_analysis['purpose_group'] = df_credit_analysis['purpose'].apply(purpose_group)

In [82]:
df_credit_analysis['purpose_group'].sort_values().unique()

array(['car', 'education', 'house', 'property', 'real_estate', 'wedding'],
      dtype=object)

In [83]:
df_credit_analysis[df_credit_analysis['purpose_group'].isna()]
df_credit_analysis.info()
df_credit_analysis.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21453 non-null  int64 
 1   days_employed     21453 non-null  int64 
 2   dob_years         21453 non-null  int64 
 3   education         21453 non-null  object
 4   education_id      21453 non-null  int64 
 5   family_status     21453 non-null  object
 6   family_status_id  21453 non-null  int64 
 7   gender            21453 non-null  object
 8   income_type       21453 non-null  object
 9   debt              21453 non-null  int64 
 10  total_income      21453 non-null  int64 
 11  purpose           21453 non-null  object
 12  dob_years_group   21453 non-null  object
 13  purpose_group     21453 non-null  object
dtypes: int64(7), object(7)
memory usage: 2.3+ MB


children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0           2194         53  secondary education             1   
5         0            926         27    bachelor's degree             0   
6         0           2879         43    bachelor's degree             0   
7         0            152         50  secondary education             1   
8         2           6929         35    bachelor's degree             0   
9         0           2188         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   
5  civil partnership                 1      M    business     0         40922   
6            married                 0      F    business     0         38484   
7            married                 0      M    employee     0         21731   
8  civil partnership                 1      F    employee     0         15337   
9            married                 0      M    employee     0         23108   

                               purpose dob_years_group purpose_group  
0                purchase of the house           40-49         house  
1                         car purchase           30-39           car  
2                purchase of the house           30-39         house  
3              supplementary education           30-39     education  
4                    to have a wedding           50-59       wedding  
5                purchase of the house           20-29         house  
6                 housing transactions           40-49         house  
7                            education           50-59     education  
8                     having a wedding           30-39       wedding  
9  purchase of the house for my family           40-49         house

In [84]:
df_credit_analysis['total_income'].sort_values(ascending=False)

12389    362496
19547    352136
9159     276204
20741    274402
17136    273809
          ...  
14246      3503
1598       3471
16137      3418
12982      3392
14554      3306
Name: total_income, Length: 21453, dtype: int64

In [85]:
df_credit_analysis['total_income'].describe()

count     21453.000000
mean      26450.314921
std       15710.264006
min        3306.000000
25%       17219.000000
50%       22815.000000
75%       31331.000000
max      362496.000000
Name: total_income, dtype: float64

<div class="alert alert-block alert-info">Based on the statistical description above for the column 'total_income' the minimum value is 3300 and the maximum is 362496. So we will create a level of 'total_income' with a range of 80000, with the lowest level being very_low. This is to make it easier to analyze the relationship between total_income and timely repayment.<a class="tocSkip"></a></div>

In [86]:
# function that performs categorizing into different numeric groups based on a range
def total_income_level(range_income):
    if range_income < 80000:
        return 'very_low'
    elif range_income <= 160000:
        return 'low'
    elif range_income <= 240000:
        return 'medium'
    elif range_income <= 320000:
        return 'high'
    elif range_income > 320000:
        return 'very_high'
    
    
print(total_income_level(230000))


medium


In [87]:
df_credit_analysis['total_income_level'] = df_credit_analysis['total_income'].apply(total_income_level)

In [88]:
df_credit_analysis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   children            21453 non-null  int64 
 1   days_employed       21453 non-null  int64 
 2   dob_years           21453 non-null  int64 
 3   education           21453 non-null  object
 4   education_id        21453 non-null  int64 
 5   family_status       21453 non-null  object
 6   family_status_id    21453 non-null  int64 
 7   gender              21453 non-null  object
 8   income_type         21453 non-null  object
 9   debt                21453 non-null  int64 
 10  total_income        21453 non-null  int64 
 11  purpose             21453 non-null  object
 12  dob_years_group     21453 non-null  object
 13  purpose_group       21453 non-null  object
 14  total_income_level  21453 non-null  object
dtypes: int64(7), object(8)
memory usage: 2.5+ MB


In [89]:
df_credit_analysis.groupby('total_income_level')['total_income_level'].count()

total_income_level
high             5
low            197
medium          18
very_high        2
very_low     21231
Name: total_income_level, dtype: int64

In [90]:
df_credit_analysis.head(10)

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0           2194         53  secondary education             1   
5         0            926         27    bachelor's degree             0   
6         0           2879         43    bachelor's degree             0   
7         0            152         50  secondary education             1   
8         2           6929         35    bachelor's degree             0   
9         0           2188         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   
5  civil partnership                 1      M    business     0         40922   
6            married                 0      F    business     0         38484   
7            married                 0      M    employee     0         21731   
8  civil partnership                 1      F    employee     0         15337   
9            married                 0      M    employee     0         23108   

                               purpose dob_years_group purpose_group  \
0                purchase of the house           40-49         house   
1                         car purchase           30-39           car   
2                purchase of the house           30-39         house   
3              supplementary education           30-39     education   
4                    to have a wedding           50-59       wedding   
5                purchase of the house           20-29         house   
6                 housing transactions           40-49         house   
7                            education           50-59     education   
8                     having a wedding           30-39       wedding   
9  purchase of the house for my family           40-49         house   

  total_income_level  
0           very_low  
1           very_low  
2           very_low  
3           very_low  
4           very_low  
5           very_low  
6           very_low  
7           very_low  
8           very_low  
9           very_low

## Checking the hypothesis


**Is there a correlation between having children and making payments on time?**

In [91]:
debt_by_children = df_credit_analysis.pivot_table(index='children', values='debt', aggfunc=['mean', 'sum', 'count', len])
debt_by_children

mean   sum  count    len
              debt  debt   debt   debt
children                              
0         0.075444  1063  14090  14090
1         0.091658   445   4855   4855
2         0.094542   194   2052   2052
3         0.081818    27    330    330
4         0.097561     4     41     41
5         0.000000     0      9      9
20        0.105263     8     76     76

In [92]:
zero_child = int(debt_by_children.loc[0,'mean'] * 100)
print('The average percentage of debt for customers with 0 children is ~',zero_child,'%')

The average percentage of debt for customers with 0 children is ~ 7 %


In [93]:
two_child = int(debt_by_children.loc[2,'mean'] * 100)
print('The average percentage of debt for customers with 2 childrens is ~',two_child,'%')

The average percentage of debt for customers with 2 childrens is ~ 9 %


In [94]:
three_child = int(debt_by_children.loc[3,'mean'] * 100)
print('The average percentage of debt for customers with 3 childrens is ~',three_child,'%')

The average percentage of debt for customers with 3 childrens is ~ 8 %


In [95]:
twenty_child = int(debt_by_children.loc[20,'mean'] * 100)
print('The average percentage of debt for customers with 20 childrens is ~',twenty_child,'%')

The average percentage of debt for customers with 20 childrens is ~ 10 %


**Conclusion**
<div class="alert alert-block alert-info">Based on the 'debt_by_children' pivot data, the average debt with customers who do not have children is ~0.07, from 1 - 2 children it increases to ~0.09. Then for 3 children it decreased to ~ 0.08, finally up to 20 children it increased again to ~ 0.1. In conclusion, the correlation between having children and making timely repayments is not strong. <a class="tocSkip"></a></div>


**Is there a correlation between family status and timely repayment?**

In [96]:
debt_by_family_status = df_credit_analysis.pivot_table(index='family_status', values='debt', aggfunc=['mean', 'sum', 'count', len])
debt_by_family_status

mean  sum  count    len
                       debt debt   debt   debt
family_status                                 
civil partnership  0.093494  388   4150   4150
divorced           0.071130   85   1195   1195
married            0.075452  931  12339  12339
unmarried          0.097509  274   2810   2810
widow / widower    0.065693   63    959    959

**Conclusion**
<div class="alert alert-block alert-info">Based on the 'debt_by_family_status' pivot, the debt for unmarried is higher than for married, and occupies the highest position compared to other statuses. For widow / widower status, the debt is the lowest.<a class="tocSkip"></a></div>


**Is there a correlation between income levels and paying back on time?**

In [97]:
debt_by_total_income = df_credit_analysis.pivot_table(index='total_income_level', values='debt', aggfunc=['mean', 'sum', 'count', len])
debt_by_total_income

mean   sum  count    len
                        debt  debt   debt   debt
total_income_level                              
high                0.000000     0      5      5
low                 0.060914    12    197    197
medium              0.055556     1     18     18
very_high           0.500000     1      2      2
very_low            0.081343  1727  21231  21231

**Conclusion**
<div class="alert alert-block alert-info">Based on the 'debt_by_total_income' pivot, the correlation between income level and timely repayment is not strong because it can be seen that the average debt from total_income_level very_low to very_high fluctuates, there is no pattern. However, it can also be concluded that the level of very_high income is the total_income range above 320,000 from two debtors, one of whom is late paying debts. <a class="tocSkip"></a></div>

**How do credit goals affect default rates?**

In [98]:
debt_by_purpose = df_credit_analysis.pivot_table(index='purpose_group', values='debt', aggfunc=['mean', 'sum', 'count', len])
debt_by_purpose

mean  sum count   len
                   debt debt  debt  debt
purpose_group                           
car            0.093590  403  4306  4306
education      0.092200  370  4013  4013
house          0.067209  256  3809  3809
property       0.074862  190  2538  2538
real_estate    0.075286  336  4463  4463
wedding        0.080034  186  2324  2324

**Conclusion**
<div class="alert alert-block alert-info">Based on the 'debt_by_purpose' pivot, it can be concluded that customers who borrow for the purpose of purchasing cars and education have higher debt than other purposes. And for the lowest debt house.<a class="tocSkip"></a></div>


# General Conclusion
<div class="alert alert-block alert-info">
In the project 'Analyzing Borrowers' Default Risk', we get unorganized data.<br>
    
We analyze the initial information problem in the data, for example.
    
- Column 'days_employed' still has a lot of negative values. There is 1 row whose value is 'NaN'. And all rows are still float values.
- Column 'education' has duplicate values, for example it contains the same meaning but different writing (Uppercase & Lowercase).
- Column 'total_income' has 1 row which has the value 'NaN' and is aligned with 'NaN' in 'days_employed'. All rows are still float values.</br>

We can handle this with several methods and functions.
<br><br>For the problem of missing and duplicate values ​​in the data and the steps to fix it.</br></br>

- Missing values ​​in the quantitative column were successfully resolved by filling in a representative value, namely the median of the column.
-Duplicate values ​​were resolved by looking for unique values ​​and categorizing them.<a class="tocSkip"></a></div>

<div class="alert alert-block alert-info"><br>From the 4 pivot tables made in the 'Checking Hypotheses' section it can be concluded:</br>
    
- The correlation between the number of children and the time of repayment of debt is not strong, but customers with 20 children are constrained to pay debts on time than customers with children under 20.
- Unmarried customers pay debts more late than married ones. For clients who are widows or widowers, they are most diligent in paying their debts on time.
- The correlation between total income and the time of repayment of debt is not strong, but customers with a total income of 120,000 to / 150,000 are the most late in paying debts than customers with a total income below 120,000 or above 150,000.
- Customers who make loans to purchase cars occupy number 1 for late paying debts, and in position 2 are customers with the aim of educational expenses.<a class="tocSkip"></a></div>
